Update `snaps.md`: Text and titles (#354)
MetaMask/metamask-docs: docs/guide/snaps.md

## MetaMask Snaps: Overview

MetaMask Snaps consists of two things: (1) a way to run untrusted JavaScript inside the MetaMask application, and (2) APIs for websites and MetaMask to communicate with individual snaps. As with MetaMask’s [Ethereum JavaScript provider API](https://docs.metamask.io/guide/ethereum-provider.html), communication occurs via JSON-RPC requests and responses.

### Current Functionality

At present, snaps can make JSON-RPC requests of MetaMask, and websites can make JSON-RPC requests of MetaMask as well as snaps running inside MetaMask. In this way, snaps can expand MetaMask’s RPC API and change the behavior of the MetaMask application at runtime. Over time, MetaMask will use this to support different blockchains and other decentralized protocols, novel kinds of cryptography and crypto assets, and many more features.

For the prototype Snaps system, snaps cannot modify the MetaMask UI, but can extend the MetaMask RPC API and exchange arbitrary messages with websites visited by the user. Therefore, the user interface for any snap must exist entirely in the website during the prototype stage.

### Execution Environment

> Snaps are untrusted JavaScript programs but execute safely inside the MetaMask application. To isolate snaps from the rest of the application and to provide a “fully virtualizable” execution environment, MetaMask uses [Secure EcmaScript (SES)](https://github.com/endojs/endo/tree/master/packages/ses), a subset of JavaScript developed by [Agoric](https://agoric.com/). Among other things, SES allows us to restrict access to global JavaScript APIs and to isolate untrusted code from other parts of the application. SES does this at the cost of some performance and incompatibility with some JavaScript practices, such as modifying prototypes of intrinsic objects, e.g., the Promise constructor.

Reviewer: Nit: Maybe better to put the aside in parentheses

```suggestion
Snaps are untrusted JavaScript programs but execute safely inside the MetaMask application. To isolate snaps from the rest of the application and to provide a “fully virtualizable” execution environment, MetaMask uses [Secure EcmaScript (SES)](https://github.com/endojs/endo/tree/master/packages/ses), a subset of JavaScript developed by [Agoric](https://agoric.com/). Among other things, SES allows us to restrict access to global JavaScript APIs and to isolate untrusted code from other parts of the application. SES does this at the cost of some performance and incompatibility with some JavaScript practices, such as modifying prototypes of intrinsic objects (e.g. the Promise constructor).
```

### Getting Flask running in your local environment

[MetaMask Flask](https://metamask.io/flask) is a distribution channel for developmental, experimental, and kooky stuff for MetaMask. Snaps is just the first feature to be released through it.

- Start by downloading [Flask](https://metamask.io/flask) and adding it to one of the available browsers of choice

After adding Flask as a browser extension, if you currently have another build of MetaMask installed in your browser, disable it (ensure you have your Secret Recovery Phrase backed up!).

### Serving a Snap to Your Local Environment

With Flask in your browser, it's time to plug something into it. If you don't want to download any code for the moment, check out the example snap [deployed here](https://metamask.github.io/snap-template/) and follow the instructions below under 'Interacting with the Snap'.

If you're ready to jump in and see how a Snap looks on the backend, follow the steps below!

Open up a new terminal window, and do the following:

- clone the snaps monorepo [here](https://github.com/MetaMask/snaps-skunkworks)

> - `cd` into the repo, and run `nvm use 14.15` our reccomended version of nodejs

Reviewer 2: Slight typo there, it should be "recommended". Can the line be worded better? Feels like the comma should be placed elsewhere.

Reviewer: ```suggestion
- `cd` into the repo, and run `nvm use 14.15` our recommended version of nodejs
```

Reviewer: How does this sound?

```suggestion
- `cd` into the repo, and then run `nvm use 14.15` to switch to our recommended version of Node.js
```

Reviewer 2: Definitely better

- navigate to the "Hello, World!" snap: `cd packages/example-snap`
- install [@metamask/snaps-cli](https://npmjs.com/package/@metamask/snaps-cli) package locally: `yarn add snaps-cli`
- run `yarn mm-snap serve`

Your terminal window should now be serving the example snap locally. Time to see the functionality in action!

### Interacting with the Snap

If you're serving the snap locally, open up a browser window, and navigate to `http://localhost:8080`.

> You should see a simple message explaining the functionality of the snap, and two buttons. Start by clicking the `Connect` button; your Flask extension should pop up and require you to click through two notifications.

Reviewer 2: Do we need the comma in the first sentence?

Reviewer: I think it helps clarify that the buttons are listed as things the user should see, rather than as being explained by the message.

Reviewer 2: Ah, that's right.
